src/main/webapp/WEB-INF/pages/product.jsp

```diff
@@ -5,27 +5,76 @@
 
 <jsp:useBean id="product" type="com.es.phoneshop.model.product.Product" scope="request"/>
 <tags:master pageTitle="Product Details">
+  <p>
+    Cart: ${cart}
+  </p>
+  <c:if test="${not empty param.add_message}">
+    <div class="success">
+        ${param.add_message}
+    </div>
+  </c:if>
+  <c:if test="${not empty error}">
+    <div class="error">
+      Item not added to card
+    </div>
+  </c:if>
   <p>
     ${product.description}
   </p>
-  <table>
+  <form method="post">
+    <table>
+        <tr>
+          <td>Image</td>
+          <td><img src="${product.imageUrl}"></td>
+        </tr>
+        <tr>
+          <td>code</td>
+          <td>${product.code}</td>
+        </tr>
       <tr>
-        <td>Image</td>
-        <td><img src="${product.imageUrl}"></td>
+        <td>stock</td>
+        <td>${product.stock}</td>
       </tr>
-      <tr>
-        <td>code</td>
-        <td>${product.code}</td>
-      </tr>
-    <tr>
-      <td>stock</td>
-      <td>${product.stock}</td>
-    </tr>
-      <tr>
-        <td>Price</td>
-        <td>
-          <fmt:formatNumber value="${product.price.currentPrice}" type="currency" currencySymbol="${product.price.currency.symbol}"/>
-        </td>
+        <tr>
+          <td>Price</td>
+          <td class="price">
+            <fmt:formatNumber value="${product.price.currentPrice}" type="currency" currencySymbol="${product.price.currency.symbol}"/>
+          </td>
+        <tr>
+          <td>quantity</td>
+          <td>
+            <input name="quantity" class="price" value="${not empty param.quantity ? param.quantity : 1}">
+            <c:if test="${not empty error}">
+              <div class="error">
+                ${error}
+              </div>
+            </c:if>
+          </td>
+        </tr>
       </tr>
-  </table>
+    </table>
+    <button>Add to card</button>
+    <c:if test="${not empty viewed.history}">
```

Comment:
it is code duplication i tried make tag out of this block<c:if> </c:if>, with viewed as attribute, but it didn't work. 
Should I keep trying this way?

Reply:
It you have time, you can try more to extract the code into a tag. If not, it's fine to have some code duplication here.